### ISSUE TITLE
MainWP incompatible with OpenSSL 3.0

### ISSUE THREAD

**Issue by lucasRolff** [mainwp/mainwp#632]
I'm a hosting provider, and we've recently started migrations from RHEL8 to RHEL9 based systems.
One of the big changes between the two, is the fact OpenSSL changed from v1.1.1 to v3.0.

We've however, found incompatibilities with MainWP's authentication process for OpenSSL 3.0, which results in authentication failures.

`openssl_verify` and `openssl_sign` in PHP by default uses `OPENSSL_ALGO_SHA1` - this algorithm isn't available by default in OpenSSL 3.0.
Since MainWP relies on the default algorithm, this causes authentication failures, since the `$signature` from `openssl_sign` will return `NULL`.

Ideally MainWP should switch to `OPENSSL_ALGO_SHA256` or higher for sign and verify operations, this will mean MainWP will work out of the box on OpenSSL 3.0 systems.

The issue has been verified on setups:
- mainWP Dashboard using v1.1.1 and MainWP child using v3.0
- mainWP Dashboard using v3.0 and MainWP child using v3.0

A temporarily solution for hosting providers is to enable legacy crypto policies, but ideally we'd want to avoid that for security reasons. This resolves the issue system wide.

It also seems like on can use `mainwp_child_openssl_sign_algo` to override it, which can be used by site owners.
What's interesting is there's a setting in "Advanced Settings" to control it, and it says the default is `OPENSSL_ALGO_SHA256`, but only if one actually saves the settings. Default installations use `OPENSSL_ALGO_SHA1`.

**Comment by bogdan-mainwp** [comment]
Hi @lucasRolff, 

We just released MainWP version 4.5 2 days ago which addresses this problem.
Have you had a chance to review the updates? 
Here you can find our KB article that describes relevant changes
https://kb.mainwp.com/docs/custom-signature-algorithm-and-phpseclib-as-alternative-to-openssl/

**Comment by lucasRolff** [comment]
Hi @bogdan-mainwp,

> We just released MainWP version 4.5 2 days ago which addresses this problem.

I don't believe 4.5 addresses the problem, since I installed MainWP 1 hour prior to making the issue to test it myself, and it indeed does not work even in 4.5.

`/wp-admin/admin.php?page=ServerInformation` also says "OpenSSL Working Status": "No"
The function `get_openssl_working_status()` does not pass the algorithm to openssl_sign/verify functions either from 4.5:
```php
        public static function get_openssl_working_status() {

                $ok = false;

                if ( function_exists( 'openssl_verify' ) && function_exists( 'openssl_pkey_new' ) ) {

                        $conf = array(
                                'private_key_bits' => 2048,
                        );

                        $conf_loc = MainWP_System_Utility::get_openssl_conf();
                        if ( ! empty( $conf_loc ) ) {
                                $conf['config'] = $conf_loc;
                        }

                        $res = openssl_pkey_new( $conf );

                        @openssl_pkey_export( $res, $privkey, null, $conf ); // phpcs:ignore -- prevent warning.
                        $details = openssl_pkey_get_details( $res );

                        if ( is_array( $details ) && isset( $details['key'] ) ) {
                                $publicKey = $details['key'];
                                $data      = 'working status';
                                openssl_sign( $data, $signature, $privkey ); // phpcs:ignore WordPress.PHP.DiscouragedPHPFunctions -- base64_encode used for http encoding compatible.
                                if ( ! empty( $signature ) ) {
                                        $ok = openssl_verify( $data, $signature, $publicKey );
                                }
                        }
                }

                return 1 === $ok;
        }
``` 

Whether upgrading from previous versions to 4.5 resolves this by "upgrading" the default algorithm to SHA256, I don't know, but a clean 4.5 install simply uses SHA1, and thus doesn't work with OpenSSL 3.0 enabled hosts, which use default crypto policies.

For the same reason, I believe the fixes you've done in 4.5 isn't sufficient considering it can be replicated out of the box on a new 4.5 installation.

**Comment by lucasRolff** [comment]
> the get_openssl_working_status() method was added to check status while tracking down the problems with OpenSSL. This method is not used when the connection is established and verification runs.

I'm aware, but it's used as a part of your ServerInformation page, so should reflect the reality, thus should be taking the selected algorithms into account as well.

As for the overall code, `openssl_sign` and `openssl_verify` will only use a different algorithm if it's defined in `mainwp_connect_signature_algo` or `mainwp_child_openssl_sign_algo` (MainWP Dashboard or MainWP Child respectively) - since these ain't in the database by default, it will mean it will default back to `OPENSSL_ALGO_SHA1`, which isn't supported.

"misconfigured openssl" should probably be changed in the change log as well to reflect the reality. The lack of OPENSSL_ALGO_SHA1 support in OpenSSL 3.0, isn't due to a misconfiguration, but due to the fact SHA1 is insecure, so by all means should not be offered.

**Comment by bogdan-mainwp** [comment]
Hi @lucasRolff,

We went through the code and made some changes.
We made sure that SHA256 is used by default, and we added additional checks in the code so SHA1 is used only in case MainWP Child is still not updated to a compatible version.

Regarding the "misconfigured openssl" statement in the changelog, this is not related to the case when there is a problem with OpenSSL v1 - v3 compatibility problems. 

We added PHPSecLib integration for cases when OpenSSL is not set, or it is really misconfigured (usually from our experience on localhost setups).
